File: backend/app/services/system_config.py
"""
Aggregate configuration for the PV system combining DB data and defaults.
"""
from __future__ import annotations

from typing import Any, Dict, List

from .panel_service import list_panels
from .battery_service import list_batteries
from .system_defaults import DEFAULT_SYSTEM_CONFIG


def _round(value: float) -> float:
    return round(value, 2)


def _aggregate_solar(panels: List[Dict[str, Any]]) -> Dict[str, Any]:
    if not panels:
        return DEFAULT_SYSTEM_CONFIG["solar"]

    total_capacity = 0.0
    total_count = 0
    total_strings = 0
    for panel in panels:
        quantity = float(panel.get("quantity") or 0)
        rated_power = float(panel.get("ratedPowerKw") or 0)
        total_capacity += quantity * rated_power
        total_count += int(quantity)
        total_strings += int(panel.get("quantity") or 0)

    primary = panels[0]
    return {
        "capacityKw": _round(total_capacity),
        "panelRatedKw": primary.get("ratedPowerKw"),
        "panelCount": total_count,
        "strings": total_strings,
        "panelEfficiencyPercent": primary.get("efficiencyPercent"),
        "panelAreaM2": primary.get("areaM2"),
        "spec": primary,
    }


def _aggregate_battery(batteries: List[Dict[str, Any]]) -> Dict[str, Any]:
    if not batteries:
        return DEFAULT_SYSTEM_CONFIG["battery"]

    total_capacity = 0.0
    total_modules = 0
    for battery in batteries:
        quantity = float(battery.get("quantity") or 0)
        capacity = float(battery.get("capacityKwh") or 0)
        total_capacity += quantity * capacity
        total_modules += int(quantity)

    primary = batteries[0]
    return {
        "capacityKwh": _round(total_capacity),
        "moduleCapacityKwh": primary.get("capacityKwh"),
        "moduleCount": total_modules,
        "maxDepthOfDischargePercent": primary.get("maxDepthOfDischargePercent"),
        "chargeRateKw": primary.get("chargeRateKw"),
        "dischargeRateKw": primary.get("dischargeRateKw"),
        "efficiencyPercent": primary.get("efficiencyPercent"),
        "spec": primary,
    }


def get_system_config() -> Dict[str, Any]:
    try:
        panels = list_panels()
        batteries = list_batteries()
    except Exception:  # pragma: no cover - DB failure fallback
        return DEFAULT_SYSTEM_CONFIG

    return {
        "location": DEFAULT_SYSTEM_CONFIG["location"],
        "solar": _aggregate_solar(panels),
        "battery": _aggregate_battery(batteries),
    }

File: src/lib/systemConfig.ts
import { BatteryConfig, SystemConfig, SolarPanelConfig } from '@/types';
import { listPanels } from './panelService';
import { listBatteries } from './batteryService';
import { DEFAULT_SYSTEM_CONFIG } from './systemDefaults';

function round(value: number): number {
  return Math.round(value * 100) / 100;
}

function aggregateSolar(panels: SolarPanelConfig[]): SystemConfig['solar'] {
  if (panels.length === 0) {
    return DEFAULT_SYSTEM_CONFIG.solar;
  }

  const totalCapacity = panels.reduce(
    (acc, panel) => acc + panel.quantity * panel.ratedPowerKw,
    0
  );
  const totalCount = panels.reduce((acc, panel) => acc + panel.quantity, 0);
  const totalStrings = panels.reduce((acc, panel) => acc + panel.quantity, 0);
  const primary = panels[0];

  return {
    capacityKw: round(totalCapacity),
    panelRatedKw: primary.ratedPowerKw,
    panelCount: totalCount,
    strings: totalStrings || undefined,
    panelEfficiencyPercent: undefined,
    panelAreaM2: undefined,
    spec: primary,
  };
}

function aggregateBattery(batteries: BatteryConfig[]): SystemConfig['battery'] {
  if (batteries.length === 0) {
    return DEFAULT_SYSTEM_CONFIG.battery;
  }

  const totalCapacity = batteries.reduce(
    (acc, battery) => acc + battery.quantity * battery.capacityKwh,
    0
  );
  const totalModules = batteries.reduce((acc, battery) => acc + battery.quantity, 0);
  const primary = batteries[0];

  return {
    capacityKwh: round(totalCapacity),
    moduleCapacityKwh: primary.capacityKwh,
    moduleCount: totalModules,
    maxDepthOfDischargePercent: undefined,
    chargeRateKw: undefined,
    dischargeRateKw: undefined,
    efficiencyPercent: undefined,
    spec: primary,
  };
}

export async function getSystemConfig(): Promise<SystemConfig> {
  try {
    const [panels, batteries] = await Promise.all([listPanels(), listBatteries()]);

    return {
      location: DEFAULT_SYSTEM_CONFIG.location,
      solar: aggregateSolar(panels),
      battery: aggregateBattery(batteries),
    };
  } catch (error) {
    console.error('No se pudo obtener la configuración del sistema. Usando valores por defecto.', error);
    return DEFAULT_SYSTEM_CONFIG;
  }
}

export { DEFAULT_SYSTEM_CONFIG };
